core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -328,10 +330,11 @@ private HttpResponse convertResponse(ServiceRequestContext ctx, HttpRequest req,
                     ((CompletionStage<?>) result)
                             .thenApply(object -> convertResponse(ctx, req, newHeaders, object,
                                                                  newTrailers))
-                            .exceptionally(cause -> exceptionHandler.handleException(ctx, req, cause)));
+                            .exceptionally(
+                                    cause -> handleExceptionWithContext(exceptionHandler, ctx, req, cause)));
         }
 
-        try {
+        try (SafeCloseable ignored = ctx.push()) {
```

Comment:
Question: Do we need to push context here?
`CompositeResponseConverterFunction#convertResponse()` push context already.
This seems to be caused when the users use RxJava? 
Then how about pushing context in ObservableResponseConverterFunction?

Reply:
Ah I forgot about `CompositeResponseConverterFunction`.
Whenever a `ResponseConverterFunctionProvider` is added, we should add the logic to push the ctx in `convertResponse` which is a bit cumbersome.
So I think we should rather push the ctx here and remove the pushing logic in `CompositeResponseConverterFunction`.

Reply:
That also sounds good.

Reply:
If we remove composite one, https://github.com/line/armeria/pull/2789/files#diff-3d3bc2932102a408e1bd85d2283bbc38R179 will be failed because of jumping thread non armeria thread, and it's http subscriber onNext/onComplete doesn't have ctx, so whenComplete at https://github.com/line/armeria/blob/master/rxjava2/src/main/java/com/linecorp/armeria/server/rxjava2/ObservableResponseConverterFunction.java#L86 won't have ctx with calling https://github.com/line/armeria/blob/master/rxjava2/src/main/java/com/linecorp/armeria/server/rxjava2/ObservableResponseConverterFunction.java#L107.

Reply:
I think it's okay not to push ctx when a user uses a different scheduler(thread). I think it's the responsibility of the user to push the ctx.
So I think we can remove the test case.

Reply:
But this means armeria does not guarantee call converter runs with ctx. For me it's good to guarantee it, but not sure...

e.g. call mysql/mongdb async
```

HttpResponse bar(){
  CompletableFuture future = new CompletableFuture();
  HttpResponse defer = HttpResponse.of(future);
  dbClient.call(result -> {
   // this code runs in db thread.
   future.complete(...);
  });
  return response;
}
```

so we need to wrap one more event loop or make context-aware to make a response converter has ctx? 
```

HttpResponse bar(){
  CompletableFuture future = RequestContext.makeContextAware(new CompletableFuture());
  HttpResponse defer = ....;
  dbClient.call(result -> {
   // this code runs in db thread.
   future.complete(...); 
  });
  return response;
}
```

Reply:
As we discussed in #2232, we cannot guarantee that the `ctx` is pushed automatically whenever it's needed.
A user might use another library(e.g. `CompletableFutures` from Spotify which we cannot make the returned future context aware) or use another thread such as the example you gave above.

If we push the ctx automatically, the first example is going to work in Annotated services.
But the same code is not going to work in other services unless a user explicitly pushes the ctx.

So it's the matter of acting consistently and being aware of which thread calls the logic. I think we don't have to take acount of pushing ctx in the thread local when another thread is used. 😄

Reply:
I can agree with that, but as a user of annotated service, we should provide a unified experience? like the converter has ctx, so we should make it in ObservableResponseConverterFunction at subscribing phase but not in annotatedService as this PR's change, because we should let user know get ctx implicitly is lucky but not 100% guarantee.

Reply:
> I think even it's 3rd libraries future wrapper, we just need to guarantee to call at request/response/error converter

What I meant is not just about the converter but about general cases. For example:
```java
CompletableFutures.allOf(webClient.get("/").aggregate(), ...).handle((combined, cause) -> {
    // A user has to push the ctx because we cannot make the future
    // as contextAwareFuture
    try (SafeClosealbe ignored = ctx.push()) {
        ...
    }
})
```

> we should provide a unified experience?

Yes, of course! We should do that. Just not sure that using another scheduler is one of the unified experiences.

Anyway, I think it's no harm to push the ctx in the composite converter, so I guess it's ok as it is. 😄

Reply:
I see. I mean I like the current changes and I can remove the ctx check converter. Should we keep it and add to non-rx converter too? If so it means armeria starts to guarantee there is a ctx in the thread when calling request/response/error converters.

Reply:
>  If so it means armeria starts to guarantee there is a ctx in the thread when calling request/response/error converters for most cases.

Yeah, that's the purpose of this PR. 😄 Let's leave it as it is.